Project1/MoveHelper.cpp

```diff
@@ -142,3 +142,52 @@ std::vector<ID> MoveHelper::getOtherAlliedCellIDs(const Unit& unit) const{
 }
 	////////////////////////////////////////////////////////////////////////////////
 
+MovementPath MoveHelper::getShortestPath(const Unit& unit, ID destination) {
+	return MovementPath(unit, _map.getShortestPathsMap(unit), _map[destination]);
+}
+
+//TODO(Torrey): optomize this
+MovementPath& MoveHelper::reRoutePath(MovementPath& path, ID destination) {
+	if (path.contains(destination)) {
+		path.trimPath(destination);
+		return path;
+	}
+	if (path.back().isAdjacent(destination)) { //Take Adjacent option if available
+		auto cost = path.back().getEdge(destination).value().getCost(path._unit.getMobility().getMobilitySet());
+		if (cost && cost.value() + path.getCost() <= path._unit.getMobility().getMove()) {
+			path.push_back(_map[destination]);
+			return path;
+		}
+	}
+	CostMap cost_map = _map.getShortestPathsMap(path._unit);
+	if (!cost_map.hasKey(destination)) { //If impossible, cut our losses
+		return path;
+	}
+	while (!path.empty()) { //Keep as much of our current path as possible
+		CostMap cost_map = _map.getShortestPathsMap(path._unit, path.back()._id, path._unit.getMobility().getMove() - path.getCost());
+		if (cost_map.hasKey(destination)) {
+			return path + MovementPath(path._unit, cost_map, _map[destination]);
+		}
+		path.pop_back();
+	}
+	Expects(false);
+	return path;
+}
+
+void MoveHelper::walkPath(Unit & unit, MovementPath path) {
+	GridCell* current_cell = _map.getCell(unit);
+	std::cout << "Start walking " << unit.getIdentity() << " from space:" << _map[unit] << " to space " << path.back()._id << std::endl;
+	for (auto& pair : path) {
+		std::cout << "Moving " << unit.getIdentity() << " from " << current_cell->_id << " to " << pair.second.get()._id << std::endl;
+		if (!unit.getMobility().canPass(_map.getUnit(pair.second))) {
+			// Walked into a stealthed unit or the like
+			//Add logic for that
+			return;
+		}
+		if (!_map.hasUnit(pair.second)) { //Skip over units we can move through
```

Comment:
We're going to need to add something to let a the units know they stepped on someone/someone stepped on them and let the moving unit know that it stepped on a tile.

Somewhat unrelatedly, what happens when someone with Pass walks through someone stealthed?

Reply:
After each step, I guess we would let every one passed through know. And stealth makes everything hard